pkg/cluster/streams.go

```diff
@@ -0,0 +1,349 @@
+		}
+	}()
+
+	desiredPatroniConfig := c.Spec.Patroni
+	if len(desiredPatroniConfig.Slots) > 0 {
+		slots = desiredPatroniConfig.Slots
+	}
+
+	// define extra logical slots for Patroni config
+	for _, stream := range c.Spec.Streams {
+		slot := map[string]string{
+			"database": stream.Database,
+			"plugin":   constants.EventStreamSourcePluginType,
+			"type":     "logical",
+		}
+		slotName := getSlotName(stream.Database, stream.ApplicationId)
+		if _, exists := slots[slotName]; !exists {
+			slots[slotName] = slot
+			publications[slotName] = stream.Tables
+		} else {
+			streamTables := publications[slotName]
+			for tableName, table := range stream.Tables {
+				if _, exists := streamTables[tableName]; !exists {
+					streamTables[tableName] = table
+				}
+			}
+			publications[slotName] = streamTables
+		}
+	}
+
+	if len(slots) > 0 {
+		desiredPatroniConfig.Slots = slots
+	} else {
+		return nil
+	}
+
+	// if streams are defined wal_level must be switched to logical
+	desiredPgParameters := map[string]string{"wal_level": "logical"}
+
+	// apply config changes in pods
+	pods, err := c.listPods()
+	if err != nil || len(pods) == 0 {
+		c.logger.Warningf("could not list pods of the statefulset: %v", err)
+	}
+	for i, pod := range pods {
+		podName := util.NameFromMeta(pods[i].ObjectMeta)
+		effectivePatroniConfig, effectivePgParameters, err := c.patroni.GetConfig(&pod)
+		if err != nil {
+			c.logger.Warningf("could not get Postgres config from pod %s: %v", podName, err)
+			continue
+		}
+
+		_, err = c.checkAndSetGlobalPostgreSQLConfiguration(&pod, effectivePatroniConfig, desiredPatroniConfig, effectivePgParameters, desiredPgParameters)
+		if err != nil {
+			c.logger.Warningf("could not set PostgreSQL configuration options for pod %s: %v", podName, err)
+			continue
+		}
+	}
+
+	// next, create publications to each created slot
+	c.logger.Debug("syncing database publications")
+	for publication, tables := range publications {
+		// but first check for existing publications
+		dbName := slots[publication]["database"]
+		if err := c.initDbConnWithName(dbName); err != nil {
+			return fmt.Errorf("could not init database connection")
+		}
+
+		currentPublications, err := c.getPublications()
+		if err != nil {
+			return fmt.Errorf("could not get current publications: %v", err)
+		}
+
+		tableNames := make([]string, len(tables))
+		i := 0
+		for t := range tables {
+			tableName, schemaName := getTableSchema(t)
+			tableNames[i] = fmt.Sprintf("%s.%s", schemaName, tableName)
+			i++
+		}
+		sort.Strings(tableNames)
+		tableList := strings.Join(tableNames, ", ")
+
+		currentTables, exists := currentPublications[publication]
+		if !exists {
+			createPublications[publication] = tableList
+		} else if currentTables != tableList {
+			alterPublications[publication] = tableList
+		}
+
+		if len(createPublications)+len(alterPublications) == 0 {
+			return nil
+		}
+
+		for publicationName, tables := range createPublications {
```

Comment:
well, there is seemingly no way  golang-native way to combine maps, but 
```golang
for publicationName, tables := range alterPublications {
			if err = c.executeAlterPublication(publicationName, tables); err != nil {
				c.logger.Warningf("%v", err)
			}
		}
```
still may be a small function to reduce code duplication

Reply:
I see what you mean. I copied it from the syncing database to [create or alter](https://github.com/zalando/postgres-operator/blob/3ce0b1e7fa236e8d54fbf066a5fb45fa11c041fd/pkg/cluster/sync.go#L916).